File: Sources/Services/RemoteControl/RemoteControlPreferences.swift
// MARK: - RemoteControlPreferences
// UserDefaults-backed Remote Control server preferences.
// macOS 14+, Swift 5.10

import Foundation
import Observation
import Security

/// UserDefaults-backed preferences for the embedded Remote Control server.
///
/// Keys are prefixed with `vs_remote_` to avoid collisions with other services.
/// Injected via `@Environment(\.remoteControlPreferences)` -- concrete type for
/// `@Observable` tracking (same pattern as `GeneralPreferences`).
@Observable
@MainActor
final class RemoteControlPreferences {

    private let defaults = UserDefaults.standard

    private enum Keys {
        static let enabled = "vs_remote_control_enabled"
        static let port = "vs_remote_control_port"
        static let bindToLocalhost = "vs_remote_bind_localhost"
        static let bonjourEnabled = "vs_remote_bonjour_enabled"
        static let idleTimeoutMinutes = "vs_remote_idle_timeout"
        static let ngrokEnabled = "vs_remote_ngrok_enabled"
    }

    // MARK: - Preferences

    /// Whether the Remote Control HTTP/WS server is enabled. Default: `false`.
    var remoteControlEnabled: Bool {
        didSet { defaults.set(remoteControlEnabled, forKey: Keys.enabled) }
    }

    /// TCP port for the HTTP/WS server. Default: `7842`. Range: 1024...65535.
    var remoteControlPort: Int {
        didSet {
            let clamped = min(max(remoteControlPort, 1024), 65535)
            if clamped != remoteControlPort { remoteControlPort = clamped }
            defaults.set(clamped, forKey: Keys.port)
        }
    }

    /// Bind the server to localhost only (127.0.0.1).
    ///
    /// **SECURITY:** Default `false` (LAN access enabled for phone control).
    /// When set to `true`, the server is unreachable from other devices on the
    /// LAN -- only local connections and SSH tunnels can reach it.
    var bindToLocalhost: Bool {
        didSet { defaults.set(bindToLocalhost, forKey: Keys.bindToLocalhost) }
    }

    /// Whether Bonjour (`_vibestudio._tcp`) service advertisement is enabled.
    ///
    /// **SECURITY:** Default `false`. Bonjour broadcasts the server's presence
    /// on the local network. Only enable if `bindToLocalhost` is also `false`.
    var bonjourEnabled: Bool {
        didSet { defaults.set(bonjourEnabled, forKey: Keys.bonjourEnabled) }
    }

    /// Whether ngrok tunnel is enabled for remote access outside LAN.
    ///
    /// **SECURITY:** Default `false`. When enabled, the local server is exposed
    /// to the internet via ngrok. PIN authentication still required.
    var ngrokEnabled: Bool {
        didSet { defaults.set(ngrokEnabled, forKey: Keys.ngrokEnabled) }
    }

    /// ngrok authtoken for authenticated tunnels.
    ///
    /// Required since ngrok v3. Get one for free at https://dashboard.ngrok.com/get-started/your-authtoken
    var ngrokAuthtoken: String {
        didSet { KeychainHelper.save(account: "vs_ngrok_authtoken", value: ngrokAuthtoken) }
    }

    /// Idle timeout in minutes before disconnecting inactive remote clients.
    /// Default: `30`.
    var idleTimeoutMinutes: Int {
        didSet {
            let clamped = max(idleTimeoutMinutes, 1)
            if clamped != idleTimeoutMinutes { idleTimeoutMinutes = clamped }
            defaults.set(clamped, forKey: Keys.idleTimeoutMinutes)
        }
    }

    // MARK: - Init

    init() {
        // remoteControlEnabled defaults to false on first launch (SECURITY: opt-in only)
        remoteControlEnabled = defaults.object(forKey: Keys.enabled) == nil
            ? false
            : defaults.bool(forKey: Keys.enabled)

        // remoteControlPort defaults to 7842
        remoteControlPort = defaults.object(forKey: Keys.port) == nil
            ? 7842
            : defaults.integer(forKey: Keys.port)

        // bindToLocalhost defaults to false — Remote Control is for phone access over LAN
        bindToLocalhost = defaults.object(forKey: Keys.bindToLocalhost) == nil
            ? false
            : defaults.bool(forKey: Keys.bindToLocalhost)

        // bonjourEnabled defaults to false (SECURITY: off by default)
        bonjourEnabled = defaults.object(forKey: Keys.bonjourEnabled) == nil
            ? false
            : defaults.bool(forKey: Keys.bonjourEnabled)

        // ngrokEnabled defaults to false (SECURITY: opt-in only)
        ngrokEnabled = defaults.object(forKey: Keys.ngrokEnabled) == nil
            ? false
            : defaults.bool(forKey: Keys.ngrokEnabled)

        // Migrate from UserDefaults to Keychain (one-time).
        // SEC-L3: only delete the UserDefaults copy AFTER Keychain confirms
        // the write — otherwise a crash between the two ops could leave the
        // token in plaintext UserDefaults forever.
        if let legacyToken = defaults.string(forKey: "vs_remote_ngrok_authtoken"), !legacyToken.isEmpty {
            let saved = KeychainHelper.save(account: "vs_ngrok_authtoken", value: legacyToken)
            if saved {
                defaults.removeObject(forKey: "vs_remote_ngrok_authtoken")
            }
            ngrokAuthtoken = legacyToken
        } else {
            ngrokAuthtoken = KeychainHelper.load(account: "vs_ngrok_authtoken") ?? ""
        }

        // idleTimeoutMinutes defaults to 30
        idleTimeoutMinutes = defaults.object(forKey: Keys.idleTimeoutMinutes) == nil
            ? 30
            : defaults.integer(forKey: Keys.idleTimeoutMinutes)
    }
}

File: shared/src/commonMain/kotlin/studio/vibe/shared/preferences/RemoteControlPreferences.kt
package studio.vibe.shared.preferences

import studio.vibe.shared.contract.CredentialStorage
import studio.vibe.shared.contract.SettingsStorage

/**
 * Key-value backed preferences for the embedded Remote Control server.
 *
 * Keys are prefixed with `vs_remote_` to avoid collisions with other services.
 * Backed by [SettingsStorage] for regular preferences and [CredentialStorage]
 * for the ngrok authtoken (secure storage — Keychain on macOS/iOS, Android Keystore, etc.).
 */
class RemoteControlPreferences(
    private val storage: SettingsStorage,
    private val credentialStorage: CredentialStorage,
) : RemoteControlPreferencesWriting {
    private object Keys {
        const val ENABLED = "vs_remote_control_enabled"
        const val PORT = "vs_remote_control_port"
        const val BIND_TO_LOCALHOST = "vs_remote_bind_localhost"
        const val BONJOUR_ENABLED = "vs_remote_bonjour_enabled"
        const val IDLE_TIMEOUT_MINUTES = "vs_remote_idle_timeout"
        const val NGROK_ENABLED = "vs_remote_ngrok_enabled"
    }

    private object CredentialKeys {
        const val NGROK_AUTHTOKEN = "vs_ngrok_authtoken"
    }

    /**
     * Whether the Remote Control HTTP/WS server is enabled. Default: `false`.
     *
     * SECURITY: Opt-in only — disabled on first launch.
     */
    override var remoteControlEnabled: Boolean
        get() = storage.getString(Keys.ENABLED)?.toBooleanStrictOrNull() ?: false
        set(value) { storage.setBool(Keys.ENABLED, value) }

    /**
     * TCP port for the HTTP/WS server. Default: `7842`. Range: 1024..65535.
     */
    override var remoteControlPort: Int
        get() = storage.getInt(Keys.PORT) ?: 7842
        set(value) {
            val clamped = value.coerceIn(1024, 65535)
            storage.setInt(Keys.PORT, clamped)
        }

    /**
     * Bind the server to localhost only (127.0.0.1). Default: `false`.
     *
     * SECURITY: Default `false` — LAN access enabled for phone control use case.
     * When `true`, only local connections and SSH tunnels can reach the server.
     */
    override var bindToLocalhost: Boolean
        get() = storage.getString(Keys.BIND_TO_LOCALHOST)?.toBooleanStrictOrNull() ?: false
        set(value) { storage.setBool(Keys.BIND_TO_LOCALHOST, value) }

    /**
     * Whether Bonjour (`_vibestudio._tcp`) service advertisement is enabled. Default: `false`.
     *
     * SECURITY: Off by default. Bonjour broadcasts the server's presence on the local
     * network. Only enable if [bindToLocalhost] is also `false`.
     */
    override var bonjourEnabled: Boolean
        get() = storage.getString(Keys.BONJOUR_ENABLED)?.toBooleanStrictOrNull() ?: false
        set(value) { storage.setBool(Keys.BONJOUR_ENABLED, value) }

    /**
     * Whether ngrok tunnel is enabled for remote access outside the LAN. Default: `false`.
     *
     * SECURITY: Off by default. When enabled, the local server is exposed to the internet
     * via ngrok. PIN authentication is still required.
     */
    override var ngrokEnabled: Boolean
        get() = storage.getString(Keys.NGROK_ENABLED)?.toBooleanStrictOrNull() ?: false
        set(value) { storage.setBool(Keys.NGROK_ENABLED, value) }

    /**
     * Idle timeout in minutes before disconnecting inactive remote clients. Default: `30`.
     */
    override var idleTimeoutMinutes: Int
        get() = storage.getInt(Keys.IDLE_TIMEOUT_MINUTES) ?: 30
        set(value) {
            val clamped = maxOf(value, 1)
            storage.setInt(Keys.IDLE_TIMEOUT_MINUTES, clamped)
        }

    /**
     * Load the ngrok authtoken from secure storage. Returns empty string if not set.
     *
     * This is a suspend function because secure storage access may be asynchronous.
     */
    override suspend fun loadNgrokAuthtoken(): String =
        credentialStorage.load(CredentialKeys.NGROK_AUTHTOKEN) ?: ""

    /**
     * Persist the ngrok authtoken to secure storage.
     */
    override suspend fun saveNgrokAuthtoken(token: String) {
        credentialStorage.save(CredentialKeys.NGROK_AUTHTOKEN, token)
    }
}
